src/lib/y2partitioner/dialogs/partition_size.rb

```diff
@@ -38,114 +147,191 @@ def initialize(disk, ptemplate, slots)
+
+        def initialize(slots)
+          textdomain "storage"
+          @slots = slots
+          largest_region = @slots.map(&:region).max_by(&:size)
+          @size = @max_size = largest_region.size
+          @min_size = Y2Storage::DiskSize.new(1)
+        end
+
+        # @return [Y2Storage::Region] of the smallest slot
+        #   that can contain the chosen size
+        def parent_region
+          regions = @slots.map(&:region)
+          suitable_rs = regions.find_all { |r| r.size >= size }
+          suitable_rs.min_by(&:size)
+        end
+
+        # @return [Y2Storage::Region] create it in the smallest slot
+        #   that can contain the chosen size
+        def region
+          parent = parent_region
+          bsize = parent.block_size
+          length = (size.to_i / bsize.to_i.to_f).ceil
+          Y2Storage::Region.create(parent.start, length, bsize)
+        end
+
+        def label
+          _("Size")
+        end
+
+        def init
+          self.value = size
+        end
+
+        def store
+          self.size = value
+        end
+
+        def validate
+          v = value
+          if v.nil? || v > max_size
+            min_s = min_size.human_ceil
+            max_s = max_size.human_floor
+            Yast::Popup.Error(
+              Yast::Builtins.sformat(
+                # error popup, %1 and %2 are replaced by sizes
+                _("The size entered is invalid. Enter a size between %1 and %2."),
+                min_s, max_s
               )
             )
-          )
+            # TODO: Let CWM set the focus
+            Yast::UI.SetFocus(Id(widget_id))
+            false
+          else
+            true
+          end
+        end
+
+        # @return [Y2Storage::DiskSize,nil]
+        def value
+          Y2Storage::DiskSize.from_human_string(super)
+        rescue ArgumentError
+          nil
+        end
+
+        # @param v [Y2Storage::DiskSize]
+        def value=(v)
+          super(v.human_floor)
+        end
+      end
+
+      # Specify start+end of the region
+      class CustomRegion < CWM::CustomWidget
+        def initialize(slots)
+          raise ArgumentError if slots.empty?
+          textdomain "storage"
+          @regions = slots.map(&:region)
+
+          largest_region = @regions.max_by(&:size)
+          self.start_block = largest_region.start
+          self.end_block = largest_region.last
         end
 
+        attr_accessor :start_block, :end_block
+
         def contents
-          contents = HVSquash(
-            # frame heading
-            term(
-              :FrameWithMarginBox,
-              _("New Partition Size"),
-              RadioButtonGroup(
-                Id(:size),
-                VBox(
-                  max_size_term,
-                  manual_size_term,
-                  manual_region_term
-                )
-              )
+          min_block = @regions.map(&:start).min
+          # FIXME: libyui widget overflow :-(
```

Comment:
I have same problem when display size in table. In the end I have to do rounding.